### DISCUSSION TITLE
Android Data

### DISCUSSION THREAD

**Discussion by PRAEst-76** [exult/exult#316] [General]
I started playing U7 using the Exult Android port on my tablet. I'd hoped to perhaps continue it on my desktop by copying over the saves (assume they were compatible) however I can't find for the life of me where Exult for Android stores it's data on Android 12.

It's not in some corner of the OS not accessable on non-rooted devices is it?

**Comment by DominusExult** [comment]
@ceckak can you help there? (and yes, the savegames are compatible)

**Reply by ceckak** [reply]
Hi @PRAEst-76,

It gets saved as application data which does end up being hard to get access to on a non-rooted device.  That's been on my TODO list to improve, but I haven't had a lot of time to spend on this recently.  In the mean time, one option is to use `adb` to make a backup of the app, and the saves will be bundled in the backup.  Extracting them is a little bit cryptic, but doable.  Here's a quick summary of the process:

```
$ adb backup info.exult
WARNING: adb backup is deprecated and may be removed in a future release
Now unlock your device and confirm the backup operation...
$ ( printf "\x1f\x8b\x08\x00\x00\x00\x00\x00" ; tail -c +25 backup.ab ) |  tar xfvz -
x apps/info.exult/_manifest
x apps/info.exult/r/app_webview
...
x apps/info.exult/sp/content.game.forgeofvirtue.xml
x apps/info.exult/sp/content.game.silverseed.xml: (null)
$ find . -name "*.sav"
./apps/info.exult/f/silverseed/mods/sifixes/saves/exult532si.sav
./apps/info.exult/f/silverseed/mods/sifixes/saves/exult577si.sav
...
```
Let me know if any of this doesn't make sense or doesn't work and I'll help you figure it out.

Ken

**Reply by PRAEst-76** [reply]
Actually, that worked perfectly. Thanks. Though it would be nice to be able to use syncthing or something to sync the saves between my tablet and laptop, however I think if you made that possible Google would update Android security to break it again. Such is their way.

**Comment by ceckak** [comment]
Glad it worked for you.  I've looked a little bit at Google's [Play Game Services](https://developers.google.com/games/services/common/concepts/savedgames), which can upload savegames to the cloud.  I've not dug into it in detail though as it wouldn't work on Amazon's Fire tablets.  I'd still like to come up with something better than what I have now though - saving to somewhere easier to access if nothing else.

**Reply by DominusExult** [reply]
I think it would need to be some more universal webstorage that could be used across platforms for all/most ports if Exult. Maybe Dropbox or/and WebDAV.

**Reply by PRAEst-76** [reply]
> I think it would need to be some more universal webstorage that could be used across platforms for all/most ports if Exult. Maybe Dropbox or/and WebDAV.

The Dropbox integration in ScummVM seems to work quite well in that regard, and it's cross platform.

**Comment by ceckak** [comment]
syncthing looks interesting - I've not seen about it before.  It looks like it uses more of a point-to-point transfer rather than a central storage site.  Will have to read about it more.

**Reply by hornowitz** [reply]
Hi everyone,

first of all, thanks for your fantastic job porting it to Android @ceckak !
I have found a workaround using resilio-sync (can also be done with syncthing). But! It is tricky... I needed to boot to TWRP in recovery mode (just on one device). So from there I was able to access the file "sifixes.cfg" (/data/data/info.exult/files/silverseed/mods/sifixes.cfg) and there I changed the savegame_path:
```
 <savegame_path>
 /storage/emulated/0/Android/media/info.exult/files/saves
 </savegame_path>
```

Then make sure that you have made this directory on your Android device. This can be accessed without being root and is the official place where you should put this kind of data.

When you have done this, you can backup the data with adb, like @ceckak said:
`adb backup -apk info.exult -f exult.backup`

And then restore it on the other device (the sifixes.cfg will be put there as well then):
`adb restore exult.backup`

And then use your favorite sync tool (rsl-sync, syncthing) and you are good to go. Tested this with my Xiaomi phone and Fire HD tablet.


I am no developer, but if you just can edit the savegames in cfg, then there should be no need for this workaround.

Bye,
Dominic

**Comment by DuendeInexistente** [comment]
Necroing this because i got to say, I'm pretty surprised the android launcher didn't just include a tab to let you edit the cfgs directly or export/import saves. I kind of assumed there'd be one the moment I saw the design.

**Comment by ceckak** [comment]
Thanks for the suggestion.  I had a decent block of time to spend on it when I did the intial port, but have struggled to find time to work on it further since then.  My initial goal was to get the basic elements of the game to be easy to use (largely motivated by my kids' interest in playing it at the time), with intent to follow up with more flexibility.  Import/export saves seems like a high priority for a follow-on, and an easy way to edit or import/export cfg would be nice as well.  Realistically, I don't have line of site to investing significant time in it any time soon though.

**Reply by DuendeInexistente** [reply]
Just a default text area presenting the cfgs and a save button would be pretty handy. Or changing the location of the cfg and savefiles, that one could be just changing a variable in the code.

I do have to say- the android port is actually much more pleasant to use than the desktop one, as I can just point the launcher to the files and there's tabs to install common addons easily. I'd actually recomend people to play on android just because of that, as the only issue I found is text input is a bit awkward.

**Reply by DominusExult** [reply]
Better would be a page with two buttons (and explaining text)
1. change savegame location

/storage/emulated/0/Android/media/info.exult/files/saves
 </savegame_path>

2. move savegames over

(Or just one button that does both)

And change the default to that for new users?

And the button(s) only become available if there are savegames in the old (current) location.

**Comment by wench** [comment]
In general allowing editing of the whole exult.cfg in Android is probably not a good idea as many settings can't or shouldn't be changed.

**Reply by DominusExult** [reply]
Yup, that's my worry as well. And apart from this idea here, nothing in the cfg needs to be done manually anymore.

**Comment by edronne49** [comment]
Hello All,

First at all congratulation to the team for the Android version.

With Android 16 none of the workaround describe here are working (i do not want to root my phone).

Is there any chance to have an "export backups" option ? 

The option to backup to google drive would be nice, in order to secure the backups.

Additionally this export option would probably help to migrate data to a new phone.

Thanks ahead !

**Reply by DominusExult** [reply]
I might be able to tinker with this soonish. A simple export of savegames to the download folder should be doable, per game a zip. And import shouldn't be too hard as well.

**Comment by edronne49** [comment]
Yes it would be very nice. Thanks ahead 👍

**Reply by DominusExult** [reply]
Next snapshot will have a savegames tab in the launcher that allows you to export to the downloads folder (and from there you can bring it into whatever cloud you want to I guess). And also import savegames from a zip.

**Reply by Silvercloak50** [reply]
I just tried the new snapshot, importing my PC KeyRing saves into it.  Pop up said 14 saves were successfully imported, opened the game - none of the 14 saves showed up.  Tried to import a second time, still no luck.  No way to confirm without rooting my phone as I'm unable to open the protected folders. :\

Thanks for working on this!

**Reply by DominusExult** [reply]
Hmm, I might need to revisit. I don't remember if I did the mods correctly if at all
